boa_engine/src/builtins/date/mod.rs

```diff
@@ -12,45 +14,67 @@ use crate::{
     },
     string::utf16,
     symbol::WellKnownSymbols,
-    value::{IntegerOrInfinity, JsValue, PreferredType},
+    value::{IntegerOrNan, JsValue, PreferredType},
     Context, JsError, JsResult,
 };
 use boa_profiler::Profiler;
-use chrono::{prelude::*, Duration, LocalResult};
+use chrono::prelude::*;
 use std::fmt::Display;
 use tap::{Conv, Pipe};
 
-/// The number of nanoseconds in a millisecond.
-const NANOS_PER_MS: i64 = 1_000_000;
-/// The number of milliseconds in an hour.
-const MILLIS_PER_HOUR: i64 = 3_600_000;
-/// The number of milliseconds in a minute.
-const MILLIS_PER_MINUTE: i64 = 60_000;
-/// The number of milliseconds in a second.
-const MILLIS_PER_SECOND: i64 = 1000;
-
-#[inline]
-fn is_zero_or_normal_opt(value: Option<f64>) -> bool {
-    value.map_or(true, |value| value == 0f64 || value.is_normal())
+/// Extracts `Some` from an `Option<T>` or returns `NaN` if the object contains `None`.
+macro_rules! some_or_nan {
+    ($v:expr) => {
+        match $v {
+            Some(dt) => dt,
+            None => return Ok(JsValue::from(f64::NAN)),
+        }
+    };
 }
 
-macro_rules! check_normal_opt {
-    ($($v:expr),+) => {
-        $(is_zero_or_normal_opt($v.into()) &&)+ true
+/// Gets a mutable reference to the inner `Date` object of `val` and stores it on `var`, or returns
+/// a `TypeError` if `val` is not a `Date` object.
+///
+/// [spec]: https://tc39.es/ecma262/#sec-thistimevalue
+macro_rules! get_mut_date {
+    (let $var:ident = $val:expr) => {
+        let mut $var = $val
+            .as_object()
+            .map(JsObject::borrow_mut)
+            .ok_or_else(|| JsNativeError::typ().with_message("'this' is not a Date"))?;
+        let $var = $var
+            .as_date_mut()
+            .ok_or_else(|| JsNativeError::typ().with_message("'this' is not a Date"))?;
     };
 }
 
+/// Abstract operation [`thisTimeValue`][spec].
+///
+/// [spec]: https://tc39.es/ecma262/#sec-thistimevalue
 #[inline]
-fn ignore_ambiguity<T>(result: LocalResult<T>) -> Option<T> {
-    match result {
-        LocalResult::Ambiguous(v, _) | LocalResult::Single(v) => Some(v),
-        LocalResult::None => None,
-    }
+pub(super) fn this_time_value(value: &JsValue) -> JsResult<Option<NaiveDateTime>> {
+    Ok(value
+        .as_object()
+        .and_then(|obj| obj.borrow().as_date().copied())
+        .ok_or_else(|| JsNativeError::typ().with_message("'this' is not a Date"))?
+        .0)
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Date(Option<NaiveDateTime>);
 
+impl Date {
+    pub(crate) fn new(dt: Option<NaiveDateTime>) -> Self {
+        Self(dt)
+    }
+    fn as_value(&self) -> JsValue {
+        match self.0 {
+            Some(dt) => JsValue::from(dt.timestamp_millis()),
+            None => JsValue::from(f64::NAN),
+        }
```

Comment:
I would use a `map_or()` or something like that, to make this a bit more idiomatic / functional